# Video Games Sales Data
This dataset contains records of popular video games in North America, Japan, Europe and other parts of the world. Every video game in this dataset has at least 100k global sales.

**(Please note that this is an adaptation of the original Video Games Sales Data workbook created by Temilola, and I use SQL here.)**

In [52]:
import pandas as pd
sales = pd.read_csv("vgsales.csv", index_col=0)
print(sales.shape)
sales.head()

(16598, 10)


Name Platform  ...  Other_Sales Global_Sales
Rank                                     ...                          
1                   Wii Sports      Wii  ...         8.46        82.74
2            Super Mario Bros.      NES  ...         0.77        40.24
3               Mario Kart Wii      Wii  ...         3.31        35.82
4            Wii Sports Resort      Wii  ...         2.96        33.00
5     Pokemon Red/Pokemon Blue       GB  ...         1.00        31.37

[5 rows x 10 columns]

## Data Dictionary

| Column        | Explanation                                                                   |
| ------------- | ----------------------------------------------------------------------------- |
| Rank          | Ranking of overall sales                                                      |
| Name          | Name of the game                                                              |
| Platform      | Platform of the games release (i.e. PC,PS4, etc.)                             |
| Year          | Year the game was released in                                                 |
| Genre         | Genre of the game                                                             |
| Publisher     | Publisher of the game                                                         |
| NA_Sales      | Number of sales in North America (in millions)                                |
| EU_Sales      | Number of sales in Europe (in millions)                                       |
| JP_Sales      | Number of sales in Japan (in millions)                                        |
| Other_Sales   | Number of sales in other parts of the world (in millions)                     |
| Global_Sales  | Number of total sales (in millions)                                           |

In [53]:
-- Creating a table in the database I connected here.
CREATE TABLE IF NOT EXISTS video_games (
    Rank INTEGER PRIMARY KEY,              -- Assuming Rank is unique and can be used as the primary key
    Name VARCHAR(255) NOT NULL,            -- Name of the game, with a reasonable length
    Platform VARCHAR(50) NOT NULL,         -- Platform of the game's release
    Year INTEGER,                          -- Year the game was released
    Genre VARCHAR(50) NOT NULL,            -- Genre of the game
    Publisher VARCHAR(255),                -- Publisher of the game
    NA_Sales NUMERIC(10, 2),               -- Sales in North America (in millions)
    EU_Sales NUMERIC(10, 2),               -- Sales in Europe (in millions)
    JP_Sales NUMERIC(10, 2),               -- Sales in Japan (in millions)
    Other_Sales NUMERIC(10, 2),            -- Sales in other parts of the world (in millions)
    Global_Sales NUMERIC(10, 2)            -- Total global sales (in millions)
);


Empty DataFrame
Columns: []
Index: []

In [54]:
!pip install pandas sqlalchemy psycopg2

Defaulting to user installation because normal site-packages is not writeable


In [55]:
from sqlalchemy import create_engine

engine = create_engine('postgresql://postgres.pftomtalgnzdymgnkcmn:RcAhoCa0Yk1NZlsf@aws-0-eu-central-1.pooler.supabase.com:5432/postgres')
# Send data to database
sales.to_sql('video_games', engine, if_exists='replace', index=True)

598

In [56]:
SELECT *
FROM public.video_games;

Rank  ... Global_Sales
0          1  ...        82.74
1          2  ...        40.24
2          3  ...        35.82
3          4  ...        33.00
4          5  ...        31.37
...      ...  ...          ...
16593  16596  ...         0.01
16594  16597  ...         0.01
16595  16598  ...         0.01
16596  16599  ...         0.01
16597  16600  ...         0.01

[16598 rows x 11 columns]

[Source](https://www.kaggle.com/gregorut/videogamesales) of dataset.

# Analysis

First, to stick to PostPostgreSQL conventions, column names should be defined in snake case.

To rename the columns of the table `video_games`, I would:
- Retrieve the column names from PostgreSQL `information_schema.columns`.
- Afterwards, I would alter the table schema based on this names.
- Use Python to execute the queries generated on the `video_games` table since I already have access to the `engine` object.

In [57]:
-- Get the column names of the video_games table
SELECT column_name
FROM information_schema.columns
WHERE table_name = 'video_games'

In [58]:
-- Alter table and update column names to lowercase
SELECT 'ALTER TABLE video_games RENAME COLUMN "' || column_name || '" TO "' || LOWER(column_name) || '";' AS stmt
FROM columns;

stmt
0   ALTER TABLE video_games RENAME COLUMN "Rank" T...
1   ALTER TABLE video_games RENAME COLUMN "Name" T...
2   ALTER TABLE video_games RENAME COLUMN "Platfor...
3   ALTER TABLE video_games RENAME COLUMN "Year" T...
4   ALTER TABLE video_games RENAME COLUMN "Genre" ...
5   ALTER TABLE video_games RENAME COLUMN "Publish...
6   ALTER TABLE video_games RENAME COLUMN "NA_Sale...
7   ALTER TABLE video_games RENAME COLUMN "EU_Sale...
8   ALTER TABLE video_games RENAME COLUMN "JP_Sale...
9   ALTER TABLE video_games RENAME COLUMN "Other_S...
10  ALTER TABLE video_games RENAME COLUMN "Global_...

In [59]:
def query(stmt):
    with engine.connect() as connection:
        try:
            return connection.execute(stmt)
        except Exception:
            pass

queries['stmt'].apply(lambda stmt: query(stmt))

0     <sqlalchemy.engine.cursor.LegacyCursorResult o...
1     <sqlalchemy.engine.cursor.LegacyCursorResult o...
2     <sqlalchemy.engine.cursor.LegacyCursorResult o...
3     <sqlalchemy.engine.cursor.LegacyCursorResult o...
4     <sqlalchemy.engine.cursor.LegacyCursorResult o...
5     <sqlalchemy.engine.cursor.LegacyCursorResult o...
6     <sqlalchemy.engine.cursor.LegacyCursorResult o...
7     <sqlalchemy.engine.cursor.LegacyCursorResult o...
8     <sqlalchemy.engine.cursor.LegacyCursorResult o...
9     <sqlalchemy.engine.cursor.LegacyCursorResult o...
10    <sqlalchemy.engine.cursor.LegacyCursorResult o...
Name: stmt, dtype: object

## Exploring the Data

> I am curious...

- How many publishers are there in the table?

In [60]:
SELECT COUNT(DISTINCT publisher) AS publishers_cnt
FROM video_games;

publishers_cnt
0             578

Incredible! There are `578` video game publishers in this table. I was surprised to see even half of that number.

> With that many publishers, I am curious...

- How many games per publisher?

In [63]:
SELECT MIN(rank) AS rank, publisher, COUNT(name) AS games_cnt
FROM video_games
GROUP BY publisher
HAVING publisher IS NOT NULL
ORDER BY rank;

rank                    publisher  games_cnt
0        1                     Nintendo        703
1       16       Microsoft Game Studios        189
2       17         Take-Two Interactive        413
3       29  Sony Computer Entertainment        683
4       30                   Activision        975
..     ...                          ...        ...
573  16475                 Inti Creates          1
574  16506                       Takuyo          1
575  16513           Interchannel-Holon          1
576  16568                   Rain Games          1
577  16573            UIG Entertainment          1

[578 rows x 3 columns]

Though `Nintendo` (which has some of the most ranked games) has published `703` video games, there are some publishers who have published more games than Nintendo. For example, `Electronic Arts` published over 1300+ games. Yet its best game is ranked as the `78th`.

> Now, I am curious...

To understand what leverage `Nintendo` had on these publishers, I will like to see the lowest ranked game for every publisher with higher `game_cnt` than `Nintendo`.

In [66]:
SELECT MAX(rank) AS rank, publisher, COUNT(name) AS games_cnt
FROM video_games
GROUP BY publisher
HAVING COUNT(name) >= (
	SELECT COUNT(name)
	FROM video_games
	WHERE publisher = 'Nintendo'
	LIMIT 1
)
ORDER BY rank DESC;

rank                     publisher  games_cnt
0  16598                    Activision        975
1  16594                       Ubisoft        921
2  16570            Namco Bandai Games        932
3  16561  Konami Digital Entertainment        832
4  16557               Electronic Arts       1351
5  16545                      Nintendo        703
6  16355                           THQ        715

> Who is `THQ` anyways? It was not in the first 10 rows returned when I checked the most ranked publishers based on their top-ranking video games.

So, Nintendo is still the least terribly ranked publisher based on the worst ranking video game.

> Again, I am curious...

What are the genres of `Nintendo` video games?

In [70]:
SELECT DISTINCT genre
FROM video_games
WHERE publisher = 'Nintendo';

How many games per `genre`?

In [76]:
SELECT MIN(rank) AS rank, genre, COUNT(name) AS games_cnt
FROM video_games
WHERE publisher = 'Nintendo'
GROUP BY genre
HAVING genre IN (SELECT genre FROM nintendo_genres)
ORDER BY rank;

`Sports` is the genre of Nintendo's highest-ranking video game. Additionally, there appears to be a correlation between `rank` and `games_cnt`. There is a tendency for more games to be created if the ranking for that genre is high.

`Misc` is obviously a collection of other genres not important enough to be listed individually, while `Action` is an outlier.

Let's see if this is true for every games.

> Yes, I am curious...

In [80]:
SELECT MIN(rank) AS rank, genre, COUNT(name) AS games_cnt
FROM video_games
GROUP BY genre
HAVING genre IN (SELECT genre FROM nintendo_genres)
ORDER BY rank;

rank         genre  games_cnt
0      1        Sports       2346
1      2      Platform        886
2      3        Racing       1249
3      5  Role-Playing       1488
4      6        Puzzle        582
5      8          Misc       1739
6     10       Shooter       1310
7     11    Simulation        867
8     17        Action       3316
9     40      Fighting        848
10    51     Adventure       1286
11   166      Strategy        681

> 😔 No correlation??? That's surprising.

Let's examine another top-ranking publisher for comparison: `Microsoft Game Studios`.

In [81]:
SELECT MIN(rank) AS rank, genre, COUNT(name) AS games_cnt
FROM video_games
WHERE publisher = 'Microsoft Game Studios'
GROUP BY genre
HAVING genre IN (
	SELECT genre 
	FROM video_games
	WHERE publisher = 'Microsoft Game Studios'
)
ORDER BY rank;

rank         genre  games_cnt
0      16          Misc         34
1      44       Shooter         28
2     140        Sports         17
3     161        Racing         22
4     193  Role-Playing         16
5     195    Simulation         18
6     579      Strategy         18
7     789        Action         21
8     987      Fighting          5
9    2656      Platform          6
10   2874     Adventure          3
11  10672        Puzzle          1

It does appear that there is a correlation again, with `Action` games as the outlier again.

> I am still curious...

Let's do this again with another top ranking publisher: `Take-Two Interactive`

In [1]:
SELECT MIN(rank) AS rank, genre, COUNT(name) AS games_cnt
FROM video_games
WHERE publisher = 'Take-Two Interactive'
GROUP BY genre
HAVING genre IN (
	SELECT genre 
	FROM video_games
	WHERE publisher = 'Take-Two Interactive'
)
ORDER BY rank;

rank         genre  games_cnt
0      17        Action         93
1     251  Role-Playing          6
2     287          Misc         27
3     315        Sports        151
4     341        Racing         20
5     385       Shooter         65
6     418     Adventure         12
7    1071      Strategy         22
8    1958      Platform         11
9    3053    Simulation          4
10  12104        Puzzle          1
11  12559      Fighting          1

From this analysis, it is evident that ranking does not have a direct impact on the `games_cnt`.

Additionally, it is observed that the highest-ranking game for a publisher does not necessarily correlate with the overall rankings of the publisher.

For each genre, I would like to determine the highest-ranking game and its associated publisher.

In [2]:
SELECT vg.rank, vg.name, vg.genre, vg.publisher
FROM (
	SELECT rank, name, genre, publisher,
		ROW_NUMBER() OVER (PARTITION BY genre ORDER BY rank) AS row_num
	FROM video_games
	WHERE publisher IS NOT NULL
) AS vg
WHERE vg.row_num = 1
ORDER BY rank;

rank  ...             publisher
0      1  ...              Nintendo
1      2  ...              Nintendo
2      3  ...              Nintendo
3      5  ...              Nintendo
4      6  ...              Nintendo
5      8  ...              Nintendo
6     10  ...              Nintendo
7     11  ...              Nintendo
8     17  ...  Take-Two Interactive
9     40  ...              Nintendo
10    51  ...              Nintendo
11   166  ...              Nintendo

[12 rows x 4 columns]

It seems that `Nintendo` dominated nearly every `genre`, which is quite fascinating!

Before I proceed to analyze the sales data, I want to examine the platforms on which these top-ranking games (from the result set above) were released.

> My curiosity is gradually waning...

In [1]:
SELECT vg.rank, vg.name, vg.genre, vg.publisher, vg.platform 
FROM (
	SELECT rank, name, genre, publisher, platform,
		ROW_NUMBER() OVER (PARTITION BY genre ORDER BY rank) AS row_num
	FROM video_games
	WHERE publisher IS NOT NULL
) AS vg
WHERE vg.row_num = 1
ORDER BY rank;

rank                                name  ...             publisher platform
0      1                          Wii Sports  ...              Nintendo      Wii
1      2                   Super Mario Bros.  ...              Nintendo      NES
2      3                      Mario Kart Wii  ...              Nintendo      Wii
3      5            Pokemon Red/Pokemon Blue  ...              Nintendo       GB
4      6                              Tetris  ...              Nintendo       GB
5      8                            Wii Play  ...              Nintendo      Wii
6     10                           Duck Hunt  ...              Nintendo      NES
7     11                          Nintendogs  ...              Nintendo       DS
8     17                  Grand Theft Auto V  ...  Take-Two Interactive      PS3
9     40             Super Smash Bros. Brawl  ...              Nintendo      Wii
10    51  Super Mario Land 2: 6 Golden Coins  ...              Nintendo       GB
11   166                     Pokemon Stadium  ...              Nintendo      N64

[12 rows x 5 columns]

From my brief research, I discovered that these platforms showcase decades of innovation in the video game industry.

**Now to the soft part...monetary part.**

### Which of the three countries has the highest average sales?

In [4]:
SELECT name, publisher,
	AVG(na_sales) * 1e6 as avg_na_sales,
	AVG(eu_sales) * 1e6 as avg_eu_sales,
	AVG(jp_sales) * 1e6 as avg_jp_sales
FROM video_games
WHERE publisher IS NOT NULL
GROUP BY publisher, name
ORDER BY publisher DESC;

name  ... avg_jp_sales
0                            Yourself Fitness  ...          0.0
1                          Monster Strike 3DS  ...     860000.0
2                                 Wasteland 2  ...          0.0
3                                     SoniPro  ...      10000.0
4                               Toushin Toshi  ...      30000.0
...                                       ...  ...          ...
11854                          Off-Road Drive  ...          0.0
11855               Men of War: Assault Squad  ...          0.0
11856                          Panzer Tactics  ...          0.0
11857                    Boulder Dash: Rocks!  ...          0.0
11858  Pirates: Legend of the Black Buccaneer  ...          0.0

[11859 rows x 5 columns]

### How do these sales compare to other regions and the global market?

In [55]:
SELECT
	vg.year,
	vg.rank,
	vg.name,
	vg.publisher,
	vg.genre,
	vg.highest_global_sales,
	vg.highest_total_sales
FROM (
    SELECT
		name,
		genre,
		publisher,
		year,
		rank,
        MAX(global_sales) OVER (PARTITION BY publisher ORDER BY year) as highest_global_sales,
        MAX(na_sales + eu_sales + jp_sales + other_sales)
			OVER (PARTITION BY publisher ORDER BY year)
				as highest_total_sales,
        ROW_NUMBER() OVER (PARTITION BY publisher ORDER BY year DESC)
			as row_num
    FROM video_games
    WHERE publisher IS NOT NULL AND year IS NOT NULL
) as vg
WHERE vg.row_num = 1
ORDER BY vg.genre DESC, vg.publisher DESC;

It's fascinating to observe that rankings don't necessarily correlate with sales figures.

For instance, the highest-selling video game globally for `Nintendo` is not the one that holds the top rank.

In [56]:
SELECT *
FROM final_query
WHERE publisher = 'Nintendo';

year   rank  ... highest_global_sales highest_total_sales
0  2016  14971  ...                82.74               82.74

[1 rows x 7 columns]

Keep in mind that the sales figures are represented in millions. For example, `82.74` translates to `82.74 million dollars`.